icefoganalytics/elcc-data-management
# Migrate Migration Setup To Using Sequelize

# Context

Switch the current migration setup to using sequelize.
See https://sequelize.org/docs/v6/other-topics/migrations/

Apparently typescript support isn't there yet, but https://github.com/sequelize/umzug is an alternative.

Note that Knex, doesn't have first class "schema" support, while Sequelize does.
This on its own might be enough reason to switch to Sequelize.

# Implementation

Used `umzug` with a `sequelize` query instance.
Migration pipeline is:
1. When the api service boots in docker it runs `api/bin/boot-app.sh`
2. The boot script runs `npm run migrate up` with runs the `migrate` script in `api/package.json`
3. The package migrate script runs `api/bin/migrate.ts` which loads the `umzug` CLI.
4. The migrate CLI script runs `api/src/db/umzug.ts` which loads the umzug config.
5. The umzug config loads the sequelize config from `api/src/db/db-client.ts`. 

The migrations path is `api/src/db/migrations`.
The migration template is located at `api/src/db/template/sample-migration.ts`

Migrations are effectively Sequelize model configs with:
- the table names specified as literal snake_case values 
- the column names in snake_case
- all `validate` functions removed
- using literal snake_case values for the model references corresponding to the referenced models table name
- using the special `MssqlDataTypes.NOW` instead of `DataTypes.NOW` from `api/src/db/mssql-data-types.ts`. This fixes a bug where `DataTypes.NOW` does not correctly map to `GETDATE()`. The Sequelize models can, and should, still use `DataTypes.NOW`

# Examples

New: https://github.com/icefoganalytics/elcc-data-management/blob/c043fe25219db8e5cb35ac32173840de388130e0/api/src/db/migrations/2023.09.06T16.28.30.create-users-table.ts
Old: https://github.com/icefoganalytics/elcc-data-management/blob/c043fe25219db8e5cb35ac32173840de388130e0/api/src/data/migrations/001_create-users.ts

# Testing Instructions

1. Drop your database via `dev down -v`
2. Boot the app via `dev up`. This will re-create your database and run all migrations.
3. Check that the app loads and that you can log in at http://localhost:8080.
4. Check that you can list all ran migrations by going to `http://localhost:3000/api/migrate`
5. You can migrate up by `dev migrate up` or by going to `http://localhost:3000/api/migrate/up`
6. You can migrate down once via `dev migrate down` or by going to `http://localhost:3000/api/migrate/down`
7. You can roll back all migrations by `dev migrate down -- --to 0`
8. You can create a new migration via `dev migrate create -- --name test.ts`. The migration will appear in the `api/src/db/migrations/` folder.
